src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/auto_crafters/AbstractAutoCrafter.java

```diff
@@ -0,0 +1,432 @@
+import me.mrCookieSlime.Slimefun.Objects.handlers.BlockTicker;
+import me.mrCookieSlime.Slimefun.api.SlimefunItemStack;
+
+/**
+ * This is the abstract super class for our auto crafters.
+ * 
+ * @author TheBusyBiscuit
+ * 
+ * @see VanillaAutoCrafter
+ * @see EnhancedAutoCrafter
+ *
+ */
+public abstract class AbstractAutoCrafter extends SlimefunItem implements EnergyNetComponent {
+
+    /**
+     * The amount of energy consumed per crafting operation.
+     */
+    private int energyConsumed = -1;
+
+    /**
+     * The amount of energy this machine can store.
+     */
+    private int energyCapacity = -1;
+
+    /**
+     * The {@link NamespacedKey} used to store recipe data.
+     */
+    protected final NamespacedKey recipeStorageKey;
+
+    // @formatter:off
+    protected final int[] background = {
+        0, 1, 2, 3, 4, 5, 6, 7, 8,
+        9, 10, 14, 15, 16, 17,
+        18, 19, 23, 25, 26,
+        27, 28, 32, 33, 34, 35,
+        36, 37, 38, 39, 40, 41, 42, 43, 44
+    };
+    // @formatter:on
+
+    @ParametersAreNonnullByDefault
+    public AbstractAutoCrafter(Category category, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
+        super(category, item, recipeType, recipe);
+
+        recipeStorageKey = new NamespacedKey(SlimefunPlugin.instance(), "recipe_key");
+
+        addItemHandler(onRightClick());
+        addItemHandler(new BlockTicker() {
+
+            @Override
+            public void tick(Block b, SlimefunItem item, Config data) {
+                AbstractAutoCrafter.this.tick(b, data);
+            }
+
+            @Override
+            public boolean isSynchronized() {
+                return true;
+            }
+        });
+    }
+
+    @Nonnull
+    private BlockUseHandler onRightClick() {
+        return e -> e.getClickedBlock().ifPresent(b -> {
+            Player p = e.getPlayer();
+
+            // Prevent blocks from being placed, food from being eaten, etc...
+            e.cancel();
+
+            if (!isValidChest(b.getRelative(BlockFace.DOWN))) {
+                SlimefunPlugin.getLocalization().sendMessage(p, "messages.auto-crafting.missing-chest");
+            } else if (SlimefunPlugin.getProtectionManager().hasPermission(p, b, ProtectableAction.INTERACT_BLOCK)) {
+                if (p.isSneaking()) {
+                    // Select a new recipe
+                    updateRecipe(b, p);
+                } else {
+                    AbstractRecipe recipe = getSelectedRecipe(b);
+
+                    if (recipe == null) {
+                        // Prompt the User to crouch
+                        SlimefunPlugin.getLocalization().sendMessage(p, "messages.auto-crafting.select-a-recipe");
+                    } else {
+                        // Show the current recipe
+                        showRecipe(p, b, recipe);
+                    }
+                }
+            }
+        });
+    }
+
+    protected void tick(@Nonnull Block b, @Nonnull Config data) {
```

Comment:
Why don't you just use `@ParametersAreNonnullByDefault`?

Reply:
For longer lines, we do.
But for such a small line of code it is fine to do it one-by-one.